### ISSUE TITLE
Inclusion in other projects shouldn't depend on CMake

### ISSUE THREAD

**Issue by forrestthewoods** [nanomsg/nanomsg#769]
I would like to make use of nanomsg in a cross-platform open source project. This is much more difficult than it should be. I should be able to include all nanomsg source and headers and have it successfully compile on each platform (Linux, Mac, Windows) just by setting the appropriate defines. 

However that isn't the case. The make process causes some source files to be excluded in some cases. Those files can't be easily controlled through the preprocessor.

I don't think there is any fundamental reason why users should be required to use cmake. To be honest I'd take it a step further and provide an SQLite-style amalgamation. It'd be fantastic if nanomsg could be included in a project on any platform simply by including a single .c file and a single .h file. There's no reason that couldn't be done.

**Comment by josephsmeng** [comment]
why don't you just right your own script for the make process then use the generated static library header file? I don't think this is nanomsg's job for this.

**Comment by forrestthewoods** [comment]
Making a library as easy to use as possible is definitely a responsibility of a library! The easier a library is to integrate into existing projects that have existing build pipelines the more the library will be used.

Look at SQLite. The amalgamated source code is literally four files. One of them 6.9 megabytes large and 197,000 lines long. But it's very, very easy to drag and drop into any existing codebase no matter what build system it uses.

Nanomsg doesn't have to go so far as to amalgamate. But it is tedious to add to an existing project. Especially if that project is cross platform. Nanomsg isn't inherently complex. It's make system isn't that complex either. But it does generate projects in such a way that makes adding nanomsg source code into an existing build pipeline unnecessarily tedious. (In my opinion.)

**Comment by Snaipe** [comment]
The simplicity you speak of with the sqlite3 build system is an illusion.

Consider why cmake was made in the first place: to provide a platform-agnostic build system whose main goal is to define a consistent lifecycle that checks, builds, and tests.

How do you test for features? How do you test for functions? How do you test for ABIs ? sqlite3 tells you to define the appropriate `-DHAVE_XXX` macro, as if it solves the underlying problem -- it doesn't, because not all features can have a fallback, and if so it just moves the problem to the user and platform maintainer. Nanomsg and almost all serious cross-platform library projects doesn't use cmake (or autotools) for show, they have to because you can't make assumptions when working with different platforms.

Okay, to be perfectly honest, I wasn't very fair with my last paragraph -- sqlite3 *do* test for features with a `./configure`, and probably generates a `config.h` that then gets bundled inside the amalgamation. However, that doesn't mean that this is still that much of a good idea ultimately, precisely because sqlite3 and nanomsg have different requirements and it just isn't fair to compare them at this level.

sqlite3 needs to be bundled in a lot of different situations, with different features enabled *within* the same platform. The traditional "one size fits all" model doesn't make sense for them, so they provide an amalgamation so that people can build their own flavor. On the other hand, nanomsg doesn't really have significant parts to disable or enable, it just does checks for platform features.

Coming back to my initial statement, the simplicity that would be provided by a nanomsg amalgamation is an illusion, because using nanomsg in any project is just as simple:

```bash
gcc foo.c -lnanomsg
```

I don't see what's wrong with the model you speak of. Just build the library with `-DNN_STATIC=ON`, and link the produced static library. You just don't have to do anything else.

**Comment by gdamore** [comment]
exactly. 

direct inclusion in SOURCE form into other projects is not a design goal. 

we can argue whether it should be. but in that case I would like to understand why this is desirable or necessary. 

testing for platform capabilities is complex.  any kind of ability to just drop in source (eg via a single C file) would either require a script to build a platform specific version of that C file or for the user to provide definitions explicitly instead of auto generating them.

**Comment by forrestthewoods** [comment]
For reference my background is video games. When often means a Windows dev environment targeting release on PC (Win/Mac/Linux) and/or console (Xbox, Playstation, Nintendo) and/or mobile (iOS, Android). Which means my line of thinking is often from a very different perspective than much of the open source community. (An often times wrong or under-informed perspective! But one sometimes worth 2 cents.)


> direct inclusion in SOURCE form into other projects is not a design goal.

Sure. My personal belief is to always include all dependencies. Tracking down dependencies can be a colossal pain in the ass. Particularly when working outside of a Linuxy ecosystem. 

For things to "just work" it's best to include everything a project needs in that project.


> Consider why cmake was made in the first place: to provide a platform-agnostic build system whose main goal is to define a consistent lifecycle that checks, builds, and tests.

I wholeheartedly agree. I'm not suggesting cmake be removed from nanomsg. But I am saying that developers who use nanomsg have their own platform-agnostic build systems which define checks, builds, and tests. A good citizen expects that it will be integrated into other systems and doesn't make that more complicated than necessary.


> any kind of ability to just drop in source (eg via a single C file) would either require a script to build a platform specific version of that C file or for the user to provide definitions explicitly instead of auto generating them.

I would most definitely NOT generate platform specific C files. I would enable users to drop in source and provide definitions explicitly. That doesn't mean force them to. Most current users of nanomsg are probably quite content with cmake. Great! But many potential users might prefer to drop source into their pre-existing build pipelines and feed in the appropriate defines.

I don't think nanomsg is far off from supporting this. It's actually quite close! To compile full source on Windows in a freshly created Visual Studio project only a few steps are required. Add a few defines, link an extra pair of windows libs, and delete poller.c.

If poller.c were appropriately wrapped with defines then all a source drop would need is defines and lib links (on windows). Files such as worker_posix.h aren't a problem as worker.c is already define switched by platform.

An amalgamated file would potentially "just work" pending other conflicts. I've not tried to produce one yet. I've also not tried to compile a full source drop on non-windows platforms. They may need one or two define switches similar to poller.c.


In summary, Nanomsg is _very_ close to supporting direct inclusion. It may also be very close to supporting amalgamation. I personally believe that both of these things are worth supporting because it enables more developers to more easily use Nanomsg on more projects.

**Comment by gdamore** [comment]
I'm going to take this under advisement for nng.  But for nanomsg, I'm not going to worry about it.  (nng has been designed to support amalgamation, and it already supports direct inclusion.)
